conductor/lib/api_client.py

```diff
@@ -59,6 +62,64 @@ def _make_request(self, verb, conductor_url, headers, params, data, raise_on_err
 #         logger.debug('response.text is: %s', response.text)
         return response
 
+    def make_prepared_request(self, verb, url, headers=None, params=None, json=None, data=None, stream=True,
+                              remove_headers_list=None, raise_on_error=True, tries=5):
+
+        """
+        Primarily used to removed enforced headers by requests.Request.
+
+        args:
+            verb: (str) of HTTP verbs
+            url: (str) url
+            headers: (dict)
+            params: (dict)
+            json: (dict)
+            data: (varies)
+            stream: (bool)
+            remove_headers_list: list of headers to remove i.e ["Transfer-Encoding"]
+            raise_on_error: (bool)
+            tries: (int) number of attempts to perform request
+
+        return: request.Response
+        """
+
+        req = requests.Request(
+            method=verb,
+            url=url,
+            headers=headers,
+            params=params,
+            json=json,
+            data=data,
+        )
+        prepped = req.prepare()
+
+        if remove_headers_list:
+            for header in remove_headers_list:
+                prepped.headers.pop(header, None)
```

Comment:
The Transfer-Encoding was still being added by requests, since the aws upload has so far never used the chunked_reader. I can certainly retry (without it) and add the stacktrace if you want.

Reply:
```
2020-06-04 14:38:55,328 conductor.lib.uploader    ERROR:  Failed to upload file: /mnt/multipart-files/35.size because of:
Traceback (most recent call last):
  File "/opt/conductor/conductor/lib/uploader.py", line 314, in do_work
    return self.do_singlepart_upload(upload, filename, file_size, md5)
  File "/opt/conductor/conductor/lib/common.py", line 269, in wrapper_function
    return orig_function(*args, **kwargs)
  File "/opt/conductor/conductor/lib/uploader.py", line 353, in do_singlepart_upload
    tries=1,
  File "/opt/conductor/conductor/lib/api_client.py", line 125, in make_prepared_request
    response.raise_for_status()
  File "/opt/conductor/python/lib/python2.7/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 501 Server Error: Not Implemented for url: https://872472272123-dev-us-west-2-render-data.s3.us-west-2.amazonaws.com/5719106386919424/data/d41d8cd98f00b204e9800998ecf8427e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA4WI3GLT56MRPGKX6%2F20200604%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20200604T143835Z&X-Amz-Expires=18000&X-Amz-SignedHeaders=content-type%3Bhost&X-Amz-Signature=91eb3537b651ce5b5cf8fe249adc16cb66a6be8b0dfa37e56ac46c5faaf6328f
2020-06-04 14:38:55,328  conductor.lib.uploader    ERROR  UploadWorker-1:  Failed to upload file: /mnt/multipart-files/35.size because of:
Traceback (most recent call last):
  File "/opt/conductor/conductor/lib/uploader.py", line 314, in do_work
    return self.do_singlepart_upload(upload, filename, file_size, md5)
  File "/opt/conductor/conductor/lib/common.py", line 269, in wrapper_function
    return orig_function(*args, **kwargs)
  File "/opt/conductor/conductor/lib/uploader.py", line 353, in do_singlepart_upload
    tries=1,
  File "/opt/conductor/conductor/lib/api_client.py", line 125, in make_prepared_request
    response.raise_for_status()
  File "/opt/conductor/python/lib/python2.7/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 501 Server Error: Not Implemented for url: https://872472272123-dev-us-west-2-render-data.s3.us-west-2.amazonaws.com/5719106386919424/data/d41d8cd98f00b204e9800998ecf8427e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA4WI3GLT56MRPGKX6%2F20200604%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20200604T143835Z&X-Amz-Expires=18000&X-Amz-SignedHeaders=content-type%3Bhost&X-Amz-Signature=91eb3537b651ce5b5cf8fe249adc16cb66a6be8b0dfa37e56ac46c5faaf6328f
2020-06-04 14:38:55,330 conductor.lib.uploader    ERROR:  ALERT! File /mnt/multipart-files/35.size retried and still failed!
expected md5 is 1B2M2Y8AsgTpgAmY7PhCfg==, real md5 is 1B2M2Y8AsgTpgAmY7PhCfg==
2020-06-04 14:38:55,330  conductor.lib.uploader    ERROR  UploadWorker-1:  ALERT! File /mnt/multipart-files/35.size retried and still failed!
expected md5 is 1B2M2Y8AsgTpgAmY7PhCfg==, real md5 is 1B2M2Y8AsgTpgAmY7PhCfg==
2020-06-04 14:38:55,330 conductor.lib.worker    ERROR:  CAUGHT EXCEPTION:
Traceback (most recent call last):
  File "/opt/conductor/conductor/lib/worker.py", line 137, in target
    output = self.do_work(job, thread_int)
  File "/opt/conductor/conductor/lib/uploader.py", line 314, in do_work
    return self.do_singlepart_upload(upload, filename, file_size, md5)
  File "/opt/conductor/conductor/lib/common.py", line 269, in wrapper_function
    return orig_function(*args, **kwargs)
  File "/opt/conductor/conductor/lib/uploader.py", line 353, in do_singlepart_upload
    tries=1,
  File "/opt/conductor/conductor/lib/api_client.py", line 125, in make_prepared_request
    response.raise_for_status()
  File "/opt/conductor/python/lib/python2.7/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 501 Server Error: Not Implemented for url: https://872472272123-dev-us-west-2-render-data.s3.us-west-2.amazonaws.com/5719106386919424/data/d41d8cd98f00b204e9800998ecf8427e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA4WI3GLT56MRPGKX6%2F20200604%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20200604T143835Z&X-Amz-Expires=18000&X-Amz-SignedHeaders=content-type%3Bhost&X-Amz-Signature=91eb3537b651ce5b5cf8fe249adc16cb66a6be8b0dfa37e56ac46c5faaf6328f
2020-06-04 14:38:55,330  conductor.lib.worker    ERROR  UploadWorker-1:  CAUGHT EXCEPTION:
Traceback (most recent call last):
  File "/opt/conductor/conductor/lib/worker.py", line 137, in target
    output = self.do_work(job, thread_int)
  File "/opt/conductor/conductor/lib/uploader.py", line 314, in do_work
    return self.do_singlepart_upload(upload, filename, file_size, md5)
  File "/opt/conductor/conductor/lib/common.py", line 269, in wrapper_function
    return orig_function(*args, **kwargs)
  File "/opt/conductor/conductor/lib/uploader.py", line 353, in do_singlepart_upload
    tries=1,
  File "/opt/conductor/conductor/lib/api_client.py", line 125, in make_prepared_request
    response.raise_for_status()
  File "/opt/conductor/python/lib/python2.7/site-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
HTTPError: 501 Server Error: Not Implemented for url: https://872472272123-dev-us-west-2-render-data.s3.us-west-2.amazonaws.com/5719106386919424/data/d41d8cd98f00b204e9800998ecf8427e?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIA4WI3GLT56MRPGKX6%2F20200604%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20200604T143835Z&X-Amz-Expires=18000&X-Amz-SignedHeaders=content-type%3Bhost&X-Amz-Signature=91eb3537b651ce5b5cf8fe249adc16cb66a6be8b0dfa37e56ac46c5faaf6328f
ERROR:root:response.text is: <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NotImplemented</Code><Message>A header you provided implies functionality that is not implemented</Message><Header>Transfer-Encoding</Header><RequestId>3587A3C1E014B007</RequestId><HostId>mfXkoiIYtx3k2VblHPVA+ADeZzhFCQKASw2j2WyAZnNf2MMohE/Zc8PYmdxX62wbcE9zK2UWUQ8=</HostId></Error>
```

Reply:
hmmm, that's not what I'm seeing.  Let's chat IRL

Reply:
Jon and I discussed IRL. It turns out that there's a bug in the `requests` library.
When attempting to upload a zero byte file (e.g.  `__init__.py`), the `requests` library mistakenly sees this non-existent body as its cue to trigger a chunked request.  (and because S3 doesn't support chunked requests, s3 returns a 501).  
Unfortunately there's no plan to fix this behavior in `requests` 2.x.   See https://github.com/psf/requests/issues/4215#issuecomment-319521235
![image](https://user-images.githubusercontent.com/10409070/83818614-a570a580-a67c-11ea-8f69-2ddbb9d8a982.png)

@jonathanrcross Can you update the docstring for the `make_prepared_request` function, describing this bug (link to `requests` issue), and justify why the `make_prepared_request` function exists.  In the future, it would be great to simplify/reduce unnecessary code. 

Thanks for digging into this with me